torchrl/objectives/cql.py

```diff
@@ -793,3 +797,330 @@ def _alpha(self):
+
+    @property
+    def in_keys(self):
+        if self._in_keys is None:
+            self._set_in_keys()
+        return self._in_keys
+
+    @in_keys.setter
+    def in_keys(self, values):
+        self._in_keys = values
+
+    @dispatch
+    def forward(self, tensordict: TensorDictBase) -> TensorDict:
+        """Computes the (DQN) CQL loss given a tensordict sampled from the replay buffer.
+
+        This function will also write a "td_error" key that can be used by prioritized replay buffers to assign
+            a priority to items in the tensordict.
+
+        Args:
+            tensordict (TensorDictBase): a tensordict with keys ["action"] and the in_keys of
+                the value network (observations, "done", "terminated", "reward" in a "next" tensordict).
+
+        Returns:
+            a tensor containing the CQL loss.
+
+        """
+        td_copy = tensordict.clone(False)
+        self.value_network(
+            td_copy,
+            params=self.value_network_params,
+        )
+
+        action = tensordict.get(self.tensor_keys.action)
+        pred_val = td_copy.get(self.tensor_keys.action_value)
+
+        if self.action_space == "categorical":
+            if action.shape != pred_val.shape:
+                # unsqueeze the action if it lacks on trailing singleton dim
+                action = action.unsqueeze(-1)
+            pred_val_index = torch.gather(pred_val, -1, index=action).squeeze(-1)
+        else:
+            action = action.to(torch.float)
+            pred_val_index = (pred_val * action).sum(-1)
+
+        cql_loss = self.cql_loss(pred_val, action)
+
+        # calculate target value
+        with torch.no_grad():
+            target_value = self.value_estimator.value_estimate(
+                td_copy,
+                target_params=self._cached_detached_target_value_params,
+            ).squeeze(-1)
+
+        with torch.no_grad():
+            priority_tensor = (pred_val_index - target_value).pow(2)
+            priority_tensor = priority_tensor.unsqueeze(-1)
+        if tensordict.device is not None:
+            priority_tensor = priority_tensor.to(tensordict.device)
+
+        tensordict.set(
+            self.tensor_keys.priority,
+            priority_tensor,
+            inplace=True,
+        )
+        loss = distance_loss(pred_val_index, target_value, self.loss_function).mean()
+
+        # add cql loss
+        loss = loss * 0.5 + cql_loss
+
+        return TensorDict({"loss": loss, "loss_cql": cql_loss.item()}, [])
+
+    @property
+    @_cache_values
+    def _cached_detached_target_value_params(self):
+        return self.target_value_network_params.detach()
+
+    def cql_loss(self, q_values, current_action):
+        """Computes the CQL loss for a batch of Q-values and one-hot encoded actions."""
+        logsumexp = torch.logsumexp(q_values, dim=-1, keepdim=True)
+        q_a = (q_values * current_action).sum(dim=-1, keepdim=True)
+
+        return (logsumexp - q_a).mean()
```

Comment:
can we return metadata too, like we're hoping to do for all losses in the future?